Assets/Scripts/Services/History/HistoryService.cs

```diff
@@ -26,5 +36,79 @@ public class HistoryService {
 			undoActionStack.Push(userAction);
 			redoActionStack.Clear();
 		}
+
+		#endregion
+		#region Routes
+		private Stack<UserAction> redoSavedRoute = new Stack<UserAction>();
+		private Stack<UserAction> undoSavedRoute = new Stack<UserAction>();
+		bool playsRoute = false;
+		public RoutesLoader routesLoader;
+		public SearchLoader searchLoader;
+		int secondsToNextRoute;
+		private HistoryController controller;
+
+		public HistoryService(HistoryController controller, int seconds, int number, string routesPath,string searchPath, string prefix = "") {
+			routesLoader = new RoutesLoader(routesPath, prefix);
+			secondsToNextRoute = seconds;
+			searchLoader = new SearchLoader(number, searchPath);
+			this.controller = controller;
+		}
+
+
+
+		public void RedoRoute() {
+			if (redoSavedRoute.Count != 0) {
+				UserAction userAction = redoSavedRoute.Pop();
+				userAction.Execute();
+				undoSavedRoute.Push(userAction);
+			}
+		}
+
+		public void UndoRoute() {
+			if (undoSavedRoute.Count != 0) {
+				UserAction userAction = undoSavedRoute.Pop();
+				userAction.UnExecute();
+				redoSavedRoute.Push(userAction);
+			}
+		}
+
+		public void loadChosenRoute(int index) {
+			redoSavedRoute = routesLoader.loadRoute(index);
+			undoSavedRoute.Clear();
+		}
+
+		public string[] getNames() {
+			return routesLoader.routesNames();
+		}
+
+		public int[] getLengths() {
+			return routesLoader.routeLengths();
+		}
+
+		public void startRoute(int index) {
+			loadChosenRoute(index);
+			controller.NetworkController.SyncRoutePlaying(true);
+		}
+
+		public IEnumerator autoRoutes() {
+			playsRoute = true;
+			while (redoSavedRoute.Count != 0 && playsRoute) {
+				RedoRoute();
+				if (redoSavedRoute.Count > 0) yield return new UnityEngine.WaitForSeconds(secondsToNextRoute);
+			}
+			playsRoute = false;
+			controller.NetworkController.SyncRoutePlaying(false);
```

Comment:
Since those events need to go through a sync before being distributed as `Action` events (on clients) there should be a single point of entry for these calls